### PR TITLE
feat: sudt transfer module

### PR DESCRIPTION
**Pull request by blckngm**
## Readme
- https://github.com/synapseweb3/crosschain-ckb-contracts/pull/27/files#r1330173718

## TODOs
```rs
fn verify_channel_input(_args: Args<'_>) -> Result<(), Error> {
    // TODO: check channel type id input.
    Ok(())
}

fn verify_packet_input(_args: Args<'_>) -> Result<(), Error> {
    // TODO: check packet type id input.
    Ok(())
}
```

### CODE REVIEW

**Comment by ashuralyk** [comment]
not passing CI test is under expectation?

File: contracts/ibc-app/sudt-transfer/src/entry.rs [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+/// (Receiving SUDT from counterparty ICS20 module.)
+///
+/// Verify the amount/type/receiver of SUDT transferred from this lock is the
+/// same as the packet data.
+///
+/// The ack message must be vec![1].
+///
+/// What about failure ack? We must verify that it's impossible to withdrawal
+/// this amount of this type of SUDT or there's a possibility of DoS. Maybe we
+/// can rely on timeout and do not explicitly acknowledge failure.
```
Comment (ashuralyk) [inline contracts/ibc-app/sudt-transfer/src/entry.rs:66]: timeout is a user-end manual setting in packet data, so if user specified a packet unable to timeout, timeout feature won't work in this case

**Comment by blckngm** [comment]
> not passing CI test is under expectation?

It's due to rust version used. I'll fix it later.

File: contracts/ibc-app/sudt-transfer/src/entry.rs (hunk 1 above)
Comment (blckngm) [reply contracts/ibc-app/sudt-transfer/src/entry.rs:66]: The counterparty module can require a timeout, e.g. in [this function](https://github.com/synapseweb3/ibc-solidity-contract/blob/1555e26cf47915afde175c821047d601324109c5/contracts/apps/20-transfer/ICS20Transfer.sol#L25).

Or we'd need to track the total amount of SUDTs sent over this channel, then we'd need a unique escrow cell or info cell (per channel per SUDT).

**Review by jjyr — APPROVED** [review]

**Comment by jjyr** [comment]
LGTM

File: contracts/ibc-app/sudt-transfer/README.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# SUDT Transfer Module for CKB IBC
+
+> :warning: **WARNING** This contract is still in the proof-of-concept stage.
```
Comment (Flouse) [inline contracts/ibc-app/sudt-transfer/README.md:3]:
```suggestion
> :warning: **WARNING** This contract is still in the proof-of-concept stage.

This smart contract (CKB lock script) provides a way to send and receive [SUDT (Simple User Defined Token)](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0025-simple-udt/0025-simple-udt.md) across different chains using the IBC protocol. The design of this module is based on the [ICS20 specification](https://github.com/cosmos/ibc/blob/main/spec/app/ics-020-fungible-token-transfer/README.md), with the goal of being compatible with the Solidity implementation of the [ICS20Transfer](https://github.com/synapseweb3/ibc-solidity-contract/blob/master/contracts/apps/20-transfer/ICS20Transfer.sol) contract.

This SUDT transfer module (`st-lock`) serves as an escrow lock, similar to [the Solidity implementation](https://github.com/synapseweb3/ibc-solidity-contract/blob/6c025378ab2640fe5b1c4ffa2a9e936659d88101/contracts/apps/20-transfer/ICS20Transfer.sol#L163). The lock's arguments should include client, channel and packet information which will be checked to ensure the security of the transfer process.

The send/recv/refund of a SUDT interchain transfer should be completed in a single CKB transaction. The specific SUDT amount in the transaction is calculated from the difference of the input/output cells. A `st-lock` transaction only allows one type of SUDT and it cannot unlock other SUDT cells.
```

Correct me if I misunderstood the code.

File: rust-toolchain [hunk 3]
```diff
@@ -0,0 +1 @@
+1.72.0
```
Comment (Flouse) [inline rust-toolchain]: Maybe update CI [RUST_TOOLCHAIN: 1.68.2](https://github.com/synapseweb3/crosschain-ckb-contracts/blob/e8e4a12bab6fb7e665f6b2556896f181ffd0b1c0/.github/workflows/ci.yaml#L14) too?

File: contracts/ibc-app/sudt-transfer/src/entry.rs [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+    output_amount: u128,
+}
+
+/// Verify and get SUDT input/output type and amount of this lock.
+///
+/// Also verifies that cell capacity doesn't change.
+fn load_sudt_io() -> Result<SudtIo, Error> {
+    let self_lock_hash = load_script_hash()?;
+
+    // Only allow a single input for now.
```
Comment (Flouse) [inline contracts/ibc-app/sudt-transfer/src/entry.rs]:
```suggestion
    // For now, allow only one input in this input group
```

File: contracts/ibc-app/sudt-transfer/src/entry.rs [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        .enumerate()
+        .filter(|(_, h)| *h == self_lock_hash);
+    let first_output_idx = outputs.next().ok_or(Error::Output)?.0;
+    // Only allow a single output for now.
+    ensure(outputs.next().is_none(), Error::Output)?;
+
+    let input_cell = load_cell(0, Source::GroupInput)?;
+    let output_cell = load_cell(first_output_idx, Source::Output)?;
+
+    let type_script_hash = load_cell_type_hash(0, Source::GroupInput)?.ok_or(Error::Input)?;
```
Comment (Flouse) [inline contracts/ibc-app/sudt-transfer/src/entry.rs:177]: Is it necessary to check the code_hash of this SUDT script?

File: rust-toolchain (hunk 3 above)
Comment (blckngm) [reply rust-toolchain]: I'd also need to update rust in the docker image.

**Comment by ashuralyk** [comment]
the CI failed because of the rust toolchain in docker imagine, is it good under your native environment ?

**Review by Flouse — APPROVED** [review]

**Pull request merged**